[GraphBolt][Doc] add tutorial and spec of OnDiskDataset (#6580)
dmlc/dgl: docs/source/graphbolt/ondisk-dataset.rst

.. _graphbolt-ondisk-dataset:

Prepare dataset
===============

> **GraphBolt** provides the ``OnDiskDataset`` class to help user organize graph

Reviewer: ```suggestion
**GraphBolt** provides the ``OnDiskDataset`` class to help you organize graph
```

> **GraphBolt** provides the ``OnDiskDataset`` class to help user organize graph
> strucutre, feature data and tasks. ``OnDiskDataset`` is designed to efficiently
> handle large graphs that do not fit into memory by storing them on disk.

Reviewer: For graphs that can fit into memory, OnDiskDataset is still the recommended way to organize it before loading to memory.
```suggestion
**GraphBolt** provides the ``OnDiskDataset`` class to help user organize plain data of graph structure, feature and tasks . ``OnDiskDataset`` is also designed to efficiently
handle large feature that do not fit into memory by storing them on disk.
```

To create an ``OnDiskDataset`` object, you need to organize all the data including
graph structure, feature data and tasks into a directory. The directory should
contain a ``metadata.yaml`` file that describes the metadata of the dataset.

Then just pass the directory path to the ``OnDiskDataset`` constructor to create
the dataset object.

.. code:: python

    from dgl.graphbolt import OnDiskDataset

    dataset = OnDiskDataset('/path/to/dataset')

The returned ``dataset`` object just loads the yaml file and does not load any
data. To load the graph structure, feature data and tasks, you need to call
the ``load`` method.

.. code:: python

    dataset.load()

The reason why we separate the ``OnDiskDataset`` object creation and data loading

> is that users may want to change some fields in the ``metadata.yaml`` file before

Reviewer: ```suggestion
is that you may want to change some fields in the ``metadata.yaml`` file before
```

> loading the data. For example, users may want to change the path of the feature

Reviewer: ```suggestion
loading the data. For example, you may want to change the path of the feature
```

Reviewer: replace all users -> you to keep the tone consistent.

data files to point to a different directory. In this case, users can just
modify the path via ``dataset.yaml_data`` directly. Then call the ``load`` method
again to load the data.

After loading the data, you can access the graph structure, feature data and
tasks through the ``graph``, ``feature`` and ``tasks`` attributes respectively.

.. code:: python

    graph = dataset.graph
    feature = dataset.feature
    tasks = dataset.tasks

The returned ``graph`` is a ``FusedCSCSamplingGraph`` object, which will be used
for sampling. The returned ``feature`` is a ``TorchBasedFeatureStore`` object,
which will be used for feature lookup. The returned ``tasks`` is a list of
``Task`` objects, which will be used for training and evaluation.

> Now let's show data folder structure and ``metadata.yaml`` file for homogeneous

Reviewer: ```suggestion
The following examples show data folder structure and ``metadata.yaml`` file for homogeneous
```

     [8. 8. 8. 8. 8. 8. 8. 8. 8. 8.]
     [9. 9. 9. 9. 9. 9. 9. 9. 9. 9.]]

For the ``tasks`` field, we have two tasks: ``node_classification`` and
``link_prediction``. For each task, we have three sets: ``train_set``,
``validation_set`` and ``test_set``.

For ``node_classification`` task, we have two fields: ``seed_nodes`` and
``labels``. The ``seed_nodes`` field is used to specify the node IDs for
training and evaluation. The ``labels`` field is used to specify the
labels. Both of them are stored in numpy files with shape of ``[num_nodes]``
like below.

.. code:: python

    train_seed_nodes.npy

    [0 1 2 3 4 5]

    train_labels.npy

    [0 1 0 1 0 1]

    val_seed_nodes.npy

    [6 7]

    val_labels.npy

    [0 1]

    test_seed_nodes.npy

    [8 9]

    test_labels.npy

    [0 1]

For ``link_prediction`` task, we have two fields: ``node_pairs``,
``negative_dsts``.  The ``node_pairs`` field is used to specify the node pairs.
The ``negative_dsts`` field is used to specify the negative destination nodes.
They are stored in numpy file with shape of ``[num_edges, 2]`` and
``[num_edges, num_neg_dsts]`` respectively like below.

.. code:: python

    train_node_pairs.npy

>     [[0 1]

Reviewer: Is it the common way to show npy data with space instead of , to separate items?

Author: I represent like this just because it's for illustration only. I don't know whether it's a common way.

Reviewer: better to keep it consistent with npy data.

Author: ok

Author: align with print np.array

    3,4
    4,5
    5,6
    6,7
    7,8
    8,9

    user_click_item.csv

    0,0
    1,1
    2,2
    3,3
    4,4
    5,5
    6,6
    7,7
    8,8
    9,9

For the feature data, we have feature data named as ``feat`` for nodes and
edges. The feature data are stored in numpy files in shape of ``[num_nodes, 10]``
and ``[num_edges, 10]`` respectively like below.

.. code:: python

    user_feat.npy

    [[0. 0. 0. 0. 0. 0. 0. 0. 0. 0.]
     [1. 1. 1. 1. 1. 1. 1. 1. 1. 1.]
     [2. 2. 2. 2. 2. 2. 2. 2. 2. 2.]
     [3. 3. 3. 3. 3. 3. 3. 3. 3. 3.]
     [4. 4. 4. 4. 4. 4. 4. 4. 4. 4.]
     [5. 5. 5. 5. 5. 5. 5. 5. 5. 5.]
     [6. 6. 6. 6. 6. 6. 6. 6. 6. 6.]
     [7. 7. 7. 7. 7. 7. 7. 7. 7. 7.]
     [8. 8. 8. 8. 8. 8. 8. 8. 8. 8.]
     [9. 9. 9. 9. 9. 9. 9. 9. 9. 9.]]

    item_feat.npy

    [[0. 0. 0. 0. 0. 0. 0. 0. 0. 0.]
     [1. 1. 1. 1. 1. 1. 1. 1. 1. 1.]
     [2. 2. 2. 2. 2. 2. 2. 2. 2. 2.]
     [3. 3. 3. 3. 3. 3. 3. 3. 3. 3.]
     [4. 4. 4. 4. 4. 4. 4. 4. 4. 4.]
     [5. 5. 5. 5. 5. 5. 5. 5. 5. 5.]
     [6. 6. 6. 6. 6. 6. 6. 6. 6. 6.]
     [7. 7. 7. 7. 7. 7. 7. 7. 7. 7.]
     [8. 8. 8. 8. 8. 8. 8. 8. 8. 8.]
     [9. 9. 9. 9. 9. 9. 9. 9. 9. 9.]]

    user_follow_user_feat.npy
      

>       [[0. 0. 0. 0. 0. 0. 0. 0. 0. 0.]

Reviewer: ```suggestion
     [[0. 0. 0. 0. 0. 0. 0. 0. 0. 0.]
```

      [1. 1. 1. 1. 1. 1. 1. 1. 1. 1.]
      [2. 2. 2. 2. 2. 2. 2. 2. 2. 2.]
      [3. 3. 3. 3. 3. 3. 3. 3. 3. 3.]
      [4. 4. 4. 4. 4. 4. 4. 4. 4. 4.]
      [5. 5. 5. 5. 5. 5. 5. 5. 5. 5.]
      [6. 6. 6. 6. 6. 6. 6. 6. 6. 6.]
      [7. 7. 7. 7. 7. 7. 7. 7. 7. 7.]
      [8. 8. 8. 8. 8. 8. 8. 8. 8. 8.]
      [9. 9. 9. 9. 9. 9. 9. 9. 9. 9.]]

    user_click_item_feat.npy
      

>       [[0. 0. 0. 0. 0. 0. 0. 0. 0. 0.]

Reviewer: ```suggestion
     [[0. 0. 0. 0. 0. 0. 0. 0. 0. 0.]
```

      homogeneous graphs. For heterogeneous graphs, it is the node type.
    - ``num``: ``int``

      The ``num`` field is used to specify the number of nodes for the node type.
      It is mandatory for both homogeneous graphs and heterogeneous graphs.

  - ``edges``: ``list``

    The ``edges`` field is used to specify the edges. It is a list of ``edge``
    objects. Each ``edge`` object has three fields: ``type``, ``format`` and
    ``path``.
    - ``type``: ``string``, optional

      The ``type`` field is used to specify the edge type. It is ``null`` for
      homogeneous graphs. For heterogeneous graphs, it is the edge type.
    - ``format``: ``string``

      The ``format`` field is used to specify the format of the edge data. It can
      only be ``csv`` for now.
    - ``path``: ``string``

      The ``path`` field is used to specify the path of the edge data. It is
      relative to the directory of ``metadata.yaml`` file.

``feature_data``
^^^^^^^^^^^^^^^

The ``feature_data`` field is used to specify the feature data. It is a list of
``feature`` objects. Each ``feature`` object has five fields: ``domain``, ``type``,

> ``name``, ``format`` and ``path``.

Reviewer: also mention metadata here?

Author: it's not ready yet. i will update the doc once it's supported.
